### ISSUE TITLE
Expression based index are handled correctly by compare_metadata

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/alembic#1106]
Trying sqlalchemy 2 beta.

Calling `compare_metadata` with a metadata that includes expression based index errors with KeyError.

To reproduce:
```py
import sqlalchemy as sa
from alembic.migration import MigrationContext
from alembic.autogenerate import compare_metadata

meta = sa.MetaData()
table = sa.Table(
    "foo",
    meta,
    sa.Column("x", sa.Integer),
    sa.Column("y", sa.Text),
    sa.Index("foo_index", sa.func.lower(sa.column('y'))),
)

engine = sa.create_engine("postgresql://scott:tiger@localhost/test")

with engine.connect() as conn:
    meta.create_all(conn)

    mc = MigrationContext.configure(connection=conn)
    compare_metadata(mc, meta)
```

Errors with the following exprection
```
Traceback (most recent call last):
  File "~\try.py", line 51, in <module>
    compare_metadata(mc, meta)
  File "my_env\lib\site-packages\alembic\autogenerate\api.py", line 134, in compare_metadata
    migration_script = produce_migrations(context, metadata)
  File "my_env\lib\site-packages\alembic\autogenerate\api.py", line 164, in produce_migrations
    compare._populate_migration_script(autogen_context, migration_script)
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 55, in _populate_migration_script
    _produce_net_changes(autogen_context, upgrade_ops)
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 89, in _produce_net_changes
    comparators.dispatch("schema", autogen_context.dialect.name)(
  File "my_env\lib\site-packages\alembic\util\langhelpers.py", line 267, in go
    fn(*arg, **kw)
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 128, in _autogen_for_tables
    _compare_tables(
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 269, in _compare_tables
    comparators.dispatch("table")(
  File "my_env\lib\site-packages\alembic\util\langhelpers.py", line 267, in go
    fn(*arg, **kw)
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 587, in _compare_indexes_and_uniques
    conn_indexes = set(  # type:ignore[assignment]
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 588, in <genexpr>
    _make_index(ix, conn_table) for ix in conn_indexes
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 285, in _make_index
    *[conn_table.c[cname] for cname in params["column_names"]],
  File "my_env\lib\site-packages\alembic\autogenerate\compare.py", line 285, in <listcomp>
    *[conn_table.c[cname] for cname in params["column_names"]],
  File "my_env\lib\site-packages\sqlalchemy\sql\base.py", line 1481, in __getitem__
    return self._index[key][1]
KeyError: None
```

**Comment by CaselIT** [comment]
This is breaking autogenerate in case for postgresql if the db contains expression indexes.

**Comment by zzzeek** [comment]
we return those now?

**Comment by CaselIT** [comment]
in pg yes for v2 we added it. I'm working on this

**Comment by CaselIT** [comment]
Ok this particular issue was fixed by ba8e22b2e29abcbfeac8a25e289191da0e7217c3

I'll try to see if we can compare expression indexes then
